implementations/rust/ockam/ockam_vault/src/software/vault_for_secure_channels/aes_aws_lc.rs

```diff
@@ -0,0 +1,107 @@
+    pub fn encrypt_message(
+        &self,
+        destination: &mut Vec<u8>,
+        msg: &[u8],
+        nonce: &[u8],
+        aad: &[u8],
+    ) -> Result<()> {
+        destination.reserve(msg.len() + TAG_LENGTH);
+        let encrypted_payload_start = destination.len();
+        destination.extend_from_slice(msg);
+
+        let tag = self
+            .encrypt_in_place_detached(
+                Nonce::try_assume_unique_for_key(nonce)
+                    .map_err(|_| VaultError::AeadAesGcmEncrypt)?,
+                aad,
+                &mut destination[encrypted_payload_start..],
+            )
+            .map_err(|_| VaultError::AeadAesGcmEncrypt)?;
+
+        destination.extend_from_slice(tag.as_ref());
+
+        Ok(())
+    }
+    pub fn decrypt_message(&self, msg: &[u8], nonce: &[u8], aad: &[u8]) -> Result<Vec<u8>> {
+        // the tag is stored at the end of the message
+        let (msg, tag) = msg.split_at(msg.len() - TAG_LENGTH);
+        let mut out = vec![0u8; msg.len()];
+        self.decrypt(
+            Nonce::try_assume_unique_for_key(nonce).map_err(|_| VaultError::AeadAesGcmDecrypt)?,
+            aad,
+            msg,
+            tag,
+            &mut out,
+        )
+        .map_err(|_| VaultError::AeadAesGcmDecrypt)?;
+
+        Ok(out)
+    }
+}
+
+cfg_if! {
+    if #[cfg(any(not(feature = "disable_default_noise_protocol"), feature = "OCKAM_XX_25519_AES256_GCM_SHA256"))] {
+        const AES_TYPE: aws_lc_rs::aead::Algorithm = aws_lc_rs::aead::AES_256_GCM;
+
+        /// This enum is necessary to be able to dispatch the encrypt or decrypt functions
+        /// based of the algorithm type. It would be avoided if `make_aes` could return existential types
+        /// but those types are not allowed in return values in Rust
+        pub struct AesGen(AeadSecret);
+
+        /// Depending on the secret type make the right type of encrypting / decrypting algorithm
+        pub(super) fn make_aes(secret: &AeadSecret) -> AesGen {
+            AesGen(secret.clone())
+        }
+    } else if #[cfg(feature = "OCKAM_XX_25519_AES128_GCM_SHA256")] {
+        const AES_TYPE: aws_lc_rs::aead::Algorithm = aws_lc_rs::aead::AES_128_GCM;
+
+        /// This enum is necessary to be able to dispatch the encrypt or decrypt functions
+        /// based of the algorithm type. It would be avoided if `make_aes` could return existential types
+        /// but those types are not allowed in return values in Rust
+        pub struct AesGen(AeadSecret);
+
+        /// Depending on the secret type make the right type of encrypting / decrypting algorithm
+        pub(super) fn make_aes(secret: &AeadSecret) -> AesGen {
+            AesGen(secret.clone())
+        }
+    }
+}
+
+impl AesGen {
+    fn encrypt_in_place_detached(
+        &self,
+        nonce: Nonce,
+        aad: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<Tag, Unspecified> {
+        let unbound_key = UnboundKey::new(&AES_TYPE, &self.0 .0).unwrap();
```

Comment:
I'm concerned if memory is properly zeroized here... We do that for our `AeadSecret` type, but that data is copied here and I don't see if `UnboundKey` or `LessSafeKey` is ever zeroized. Could you please check it?

Reply:
I've also explicit passed `zeroize` to `aes-gcm`, just in case.

Reply:
It happens inside `OPENSSL_cleanse`. ec_key is for elliptic curve keys. But anyways, it happens and that's good.